### PR TITLE
Improving the HandlesQuerySource functionality to better deal with SelectMany-GroupJoin-DefaultIfEmpty cases.

### PR DESCRIPTION
**Pull request by maumar**

### CODE REVIEW

**Comment by maumar** [comment]
@smitpatel @anpete

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/LeftOuterJoinExpression.cs [hunk 1]
```diff
@@ -5,6 +5,10 @@
 using JetBrains.Annotations;
 using Microsoft.EntityFrameworkCore.Query.Sql;
 using Microsoft.EntityFrameworkCore.Utilities;
+using Remotion.Linq.Clauses;
+using Remotion.Linq.Clauses.Expressions;
+using Remotion.Linq.Clauses.ResultOperators;
+using System.Linq;
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/LeftOuterJoinExpression.cs]: sort usings

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/SelectExpression.cs [hunk 2]
```diff
@@ -240,20 +240,11 @@ public override Expression Visit(Expression expression)
...
 
-        /// <summary>
-        ///     Determines whether or not this SelectExpression handles the given query source.
-        /// </summary>
-        /// <param name="querySource"> The query source. </param>
-        /// <returns>
-        ///     true if the supplied query source is handled by this SelectExpression; otherwise false.
-        /// </returns>
-        public virtual bool HandlesQuerySource([NotNull] IQuerySource querySource)
+        public override bool HandlesQuerySource([NotNull] IQuerySource querySource)
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/SelectExpression.cs]: Don't eat doc comments on public surface 😛

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/LeftOuterJoinExpression.cs [hunk 3]
```diff
@@ -36,6 +40,28 @@ protected override Expression Accept(ExpressionVisitor visitor)
                 : base.Accept(visitor);
         }
 
+        public override bool HandlesQuerySource(IQuerySource querySource)
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/LeftOuterJoinExpression.cs]: doc comments

**Comment by smitpatel** [comment]
looks good.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/LeftOuterJoinExpression.cs [hunk 4]
```diff
@@ -37,6 +41,35 @@ protected override Expression Accept(ExpressionVisitor visitor)
...
+        /// </returns>
+        public override bool HandlesQuerySource(IQuerySource querySource)
+        {
+            if (querySource is AdditionalFromClause additionalFromClause)
+            {
+                var subQueryModel = (additionalFromClause.FromExpression as SubQueryExpression)?.QueryModel;
+                if (subQueryModel != null
+                    && !subQueryModel.BodyClauses.Any()
+                    && subQueryModel.ResultOperators.Count == 1
+                    && subQueryModel.ResultOperators.Single() is DefaultIfEmptyResultOperator
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/LeftOuterJoinExpression.cs]: Use [0] if you've already checked length

**Comment by anpete** [comment]
@smitpatel @maumar It seems like this is still pretty ad-hoc? If so, can we move to something more general?

**Comment by maumar** [comment]
filed https://github.com/aspnet/EntityFramework/issues/7716 to track this

**Pull request merged**